### PR TITLE
Specs dict fix

### PR DESCRIPTION
**Pull request by frederikwilde**
**Context:**
In #4038 the `specs` dictionary of `QuantumScript` was customized to throw warnings when certain keys are used.
However, when using `qml.specs` on a QNode this dictionary gets copied and the result is a regular `dict` which doesn't throw warnings.

This code block doesn't throw a warning with the current state, but does throw a warning with the changes made in this PR.
```Python
import pennylane as qml

dev = qml.device('default.qubit', wires=2)

@qml.qnode(dev)
def qnode(param):
    qml.RX(param, wires=0)
    return qml.expval(qml.PauliZ(0))

qml.specs(qnode)(1.)['gate_types']
```

**Description of the Change:**
Add a `copy` method to the `SpecsDict` to make sure that also `qml.specs(qnode)` throws warning upon usage of deprecated keys.

**Benefits:**
Deprecation of specs keys are uniformly presented.

**Possible Drawbacks:**
n/a

**Related GitHub Issues:**
n/a

### CODE REVIEW

**Review by dwierichs — APPROVED** [review]
Looks good to me, thanks for catching this @frederikwilde :100: :) 
In case there is time, we could consider to remove from docstrings the usage of the deprecated keys, but that's definitely on the side and out of scope.

**Comment by frederikwilde** [comment]
You mean in the docstring of `qml.specs`? There is this [example] which still shows the deprecated keys. But I think the example should show the actual output of `qml.specs(circuit)`, even if it contains deprecated keys.

**Comment by frederikwilde** [comment]
Oh, you're referring to [this](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/transforms/specs.py#L88) docstring, right?

**Review by timmysilv — APPROVED** [review]

File: pennylane/transforms/optimization/pattern_matching.py [hunk 1]
```diff
@@ -141,10 +141,10 @@ def circuit():
     In our case, it is possible to find three CNOTs and replace this pattern with only two CNOTs and therefore
     optimizing the circuit. The number of CNOTs in the circuit is reduced by one.
 
-    >>> qml.specs(qnode)()["gate_types"]["CNOT"]
+    >>> qml.specs(qnode)()["resources"].gate_types["CNOT"]
```
Comment (Jaybsoni) [inline pennylane/transforms/optimization/pattern_matching.py:144]: So this was the old syntax which is deprecated now right?

**Review by Jaybsoni — APPROVED** [review]

File: pennylane/transforms/optimization/pattern_matching.py (hunk 1 above)
Comment (frederikwilde) [reply pennylane/transforms/optimization/pattern_matching.py:144]: Yes, with this PR `qml.specs(qnode)()["gate_types"]` gives out a deprecation warning.

**Pull request merged**